lib/util/SparqlExpressionEvaluator.js

```diff
@@ -101,121 +158,489 @@ evaluators = {
         switch (operator.resultType) {
         case 'numeric':
           // TODO: determine type instead of taking the type of the first argument
-          var type = N3Util.getLiteralType(origArgs[0]) || XSD_INTEGER;
+          var type = isNumeric(origArgs[0]) ? N3Util.getLiteralType(origArgs[0]) : XSD + 'integer';
           return '"' + result + '"^^' + type;
         case 'boolean':
           return result ? XSD_TRUE : XSD_FALSE;
+        case 'string':
+          return '"' + result + '"^^' + XSD_STRING;
         default:
           return result;
         }
       };
     })(operator, argumentExpressions);
   },
 };
-evaluators.functionCall = evaluators.operation;
 
 // Operators for each of the operator types
 operators = {
-  '+':  function (a, b) { return a  +  b; },
-  '-':  function (a, b) { return a  -  b; },
-  '*':  function (a, b) { return a  *  b; },
-  '/':  function (a, b) { return a  /  b; },
-  '=':  function (a, b) { return a === b; },
-  '!=': function (a, b) { return a !== b; },
-  '<':  function (a, b) { return a  <  b; },
-  '<=': function (a, b) { return a  <= b; },
-  '>':  function (a, b) { return a  >  b; },
-  '>=': function (a, b) { return a  >= b; },
-  '!':  function (a)    { return !a;      },
-  '&&': function (a, b) { return a &&  b; },
-  '||': function (a, b) { return a ||  b; },
-  'lang': function (a)    {
+  // 17.4.1 Functional Forms
+  '+': function (a, b) {
+    return a + b;
+  },
+  '-': function (a, b) {
+    return a - b;
+  },
+  '*': function (a, b) {
+    return a * b;
+  },
+  '/': function (a, b) {
+    return a / b;
+  },
+  '=': function (a, b) {
+    return a === b;
+  },
+  '!=': function (a, b) {
+    return a !== b;
+  },
+  '<': function (a, b) {
+    return a < b;
+  },
+  '<=': function (a, b) {
+    return a <= b;
+  },
+  '>': function (a, b) {
+    return a > b;
+  },
+  '>=': function (a, b) {
+    return a >= b;
+  },
+  '!': function (a) {
+    return !a;
+  },
+  '&&': function (a, b) {
+    return a && b;
+  },
+  '||': function (a, b) {
+    return a || b;
+  },
```

Comment:
TODO: where these inlinings necessary for the linter? Probably not, unless the linter changed. I'd inline them again, since this code now takes up a lot of space for what it does.